src/main/java/fi/metatavu/metaform/server/rest/RealmsApiImpl.java

```diff
@@ -230,17 +228,24 @@ public Response updateReply(String realmId, UUID metaformId, UUID replyId, Reply
     if (!isRealmMetaformAdmin() && !reply.getUserId().equals(loggedUserId)) {
       return createNotFound(NOT_FOUND_MESSAGE);
     }
-
-    List<String> fieldNames = new ArrayList<>(replyController.listFieldNames(reply));
     
-    for (Entry<String, Object> entry : payload.getData().entrySet()) {
+    Metaform metaformEntity = metaformTranslator.translateMetaform(metaform);
+    
+    List<String> fieldNames = new ArrayList<>(replyController.listFieldNames(reply));
+    ReplyData data = payload.getData();
+    Map<String, MetaformFieldType> fieldTypeMap = fieldController.getFieldTypeMap(metaformEntity);
+
+    for (Entry<String, Object> entry : data.entrySet()) {
       String fieldName = entry.getKey();
-      replyController.setReplyField(reply, fieldName, entry.getValue());
+      replyController.setReplyField(fieldTypeMap.get(fieldName), reply, fieldName, entry.getValue());
       fieldNames.remove(fieldName);
     }
     
     replyController.deleteReplyFields(reply, fieldNames);
     
+    // TODO: Update reply attachments
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://sonarcloud.io/organizations/metatavu-oy/rules#rule_key=squid%3AS1135))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=fi.metatavu.metaform-api%3Ametaform-api&pullRequest=47&issues=AWVk3wFz4Bfu85TPTV1E&open=AWVk3wFz4Bfu85TPTV1E)

[ISSUE_KEY]: AWVk3wFz4Bfu85TPTV1E